### PR TITLE
set -Wdocumentation for clang

### PR DESCRIPTION
**Pull request by danieljurek**
Sets -Wdocumentation for clang builds. CI is failing for this because the new parameter is detecting documentation errors in `az_http_request_builder.h`. 

@barcharcraz - While there's some repetition in the warning parameters between GCC (i.e. builds which are not using msvc or clang) I think explicitness is preferable to logic that generates the list of compiler flags. Is this a reasonable approach?

### CODE REVIEW

**Comment by barcharcraz** [comment]
separating flags by compiler is indeed the right approach. One thing to be aware of is clang-cl. I'll post a real review with the solution

**Comment by barcharcraz** [comment]
oh also, on more modern cmake versions you can do
```
target_compile_options(meow PRIVATE $<$<COMPILE_LANG_AND_ID:C,Clang>:...>)
```

But I don't think it's necessarily a good idea. Also it's not supported on all versions of cmake we decided to support

File: sdk/core/core/CMakeLists.txt [hunk 1]
```diff
@@ -45,6 +45,8 @@ add_library (
 
 if(MSVC)
   target_compile_options(az_core PRIVATE /Wall /WX /wd5031 /wd4668 /wd4820 /wd4255 /analyze)
+elseif(CMAKE_C_COMPILER_ID MATCHES "Clang")
+  target_compile_options(az_core PRIVATE -Wall -Wextra -pedantic -Werror -Wdocumentation -Wdocumentation-unknown-command -Wcast-qual)
```
Comment (barcharcraz) [inline sdk/core/core/CMakeLists.txt]: This causes a problem with clang-cl. In that case CMAKE_C_COMPILER_ID will be "Clang", however -Wall will be translated by clang-cl as /Wall, which is equivalent to -Weverything in clang (sigh).

The solution is to instead say -Xclang -Wall .....

Comment (barcharcraz) [inline sdk/core/core/CMakeLists.txt]:
```suggestion
  target_compile_options(az_core PRIVATE -Xclang -Wall -Wextra -pedantic -Werror -Wdocumentation -Wdocumentation-unknown-command -Wcast-qual)
```

File: sdk/core/core/CMakeLists.txt [hunk 2]
```diff
@@ -45,6 +45,8 @@ add_library (
 
 if(MSVC)
   target_compile_options(az_core PRIVATE /Wall /WX /wd5031 /wd4668 /wd4820 /wd4255 /analyze)
```
Comment (barcharcraz) [inline sdk/core/core/CMakeLists.txt]: /WX and -Werror are still annoying to me. you need to be able to opt out of them to support package maintainers and people using newer compilers than you.

File: sdk/keyvault/keyvault/CMakeLists.txt [hunk 3]
```diff
@@ -14,6 +14,8 @@ add_library (
 
 if(MSVC)
   target_compile_options(az_keyvault PRIVATE /Wall /WX /wd5031 /wd4668 /wd4820 /wd4255 /analyze)
+elseif(CMAKE_C_COMPILER_ID MATCHES "Clang")
+  target_compile_options(az_keyvault PRIVATE -Wall -Wextra -pedantic -Werror -Wdocumentation  -Wdocumentation-unknown-command)
```
Comment (barcharcraz) [inline sdk/keyvault/keyvault/CMakeLists.txt]:
```suggestion
  target_compile_options(az_keyvault PRIVATE -Xclang -Wall -Wextra -pedantic -Werror -Wdocumentation  -Wdocumentation-unknown-command)
```

**Comment by danieljurek** [comment]
Accepting @barcharcraz's suggestions within the scope of this issue (which is just for enabling `-Wdocumentation`). We'll expand on the usage of `-Weverything` in discussions for the other PR (https://github.com/Azure/azure-sdk-for-c/pull/261)

**Comment by barcharcraz** [comment]
My suggestion for -Xclang -Wall is within the scope of the issue. 

The PR now is `-Weverything` for `clang-cl` and `-Wall` for clang, these are _different_ warning levels.

**Comment by danieljurek** [comment]
@barcharcraz -- You are right. I re-added the -Xclang flag and that should close up on your concerns. I think this is ready to merge. If I missed something we can open up bugs. Thanks, again, for your help here.

**Pull request merged**